RyuSocks.Test/Auth/UsernameAndPasswordTests.cs

```diff
@@ -0,0 +1,182 @@
+                if (registered)
+                {
+                    Assert.True(usernameAndPassword.Authenticate(expectedUsernameAndPasswordResponse.Bytes, out _));
+                }
+                else
+                {
+                    Assert.ThrowsAny<Exception>(() =>
+                        usernameAndPassword.Authenticate(expectedUsernameAndPasswordResponse.Bytes, out _));
+                }
+            }
+
+            [Theory]
+            [InlineData(ValidUsername, ValidPassword, true)]
+            [InlineData(ValidUsername, "RandomPassword", false)]
+            [InlineData("RandomUsername", ValidPassword, false)]
+            [InlineData("RandomUsername", "RandomPassword", false)]
+            public void Authenticate_WorksOnRegisteredUserAuthenticationServerSide(string username, string password,
+                bool registered)
+            {
+                SetUsernameAndPassword(username, password);
+                UsernameAndPassword usernameAndPassword = _usernameAndPassword;
+                usernameAndPassword.Authenticate(null, out ReadOnlySpan<byte> outgoingPacket);
+                UsernameAndPasswordRequest input = new(outgoingPacket.ToArray());
+                input.Validate();
+
+                if (!registered)
+                {
+                    // Since a ByRef variable (of type ReadOnlySpan) is not accepted by a lambda expression, it is converted to a byte array
+                    byte[] outgoingPacketByte = outgoingPacket.ToArray();
+                    Assert.ThrowsAny<Exception>(() => usernameAndPassword.Authenticate(outgoingPacketByte, out _));
+                }
+                else
+                {
+                    usernameAndPassword.IsClient = false;
+                    usernameAndPassword.Authenticate(outgoingPacket, out ReadOnlySpan<byte> responsePacket);
+                    Assert.Equal(_expectedUsernameAndPasswordResponse.AsSpan(), responsePacket);
+                }
+            }
+        }
+
+        [Theory]
+        [InlineData(new byte[] { 0xFF, 0xFF, 0xAA, 0x00, 0xCC, 0xBB })]
+        public void Wrap_DoesNotModifyPaket(byte[] packet)
+        {
+            byte[] originalPacket = (byte[])packet.Clone();
+            int packetLength = packet.Length;
+            UsernameAndPassword noAuth = new();
+
+            noAuth.Wrap(packet, packetLength, null);
+
+            Assert.Equal(packetLength, packet.Length);
+            Assert.Equal(originalPacket, packet);
+        }
+
+        [Theory]
+        [InlineData(new byte[] { 0xFF, 0xFF, 0xAA, 0x00, 0xCC, 0xBB })]
+        public void Unwrap_DoesNotModifyPaket(byte[] packet)
+        {
+            byte[] originalPacket = (byte[])packet.Clone();
+            int packetLength = packet.Length;
+            UsernameAndPassword noAuth = new();
+
+            noAuth.Unwrap(packet, packetLength, out _);
+
+            Assert.Equal(packetLength, packet.Length);
+            Assert.Equal(originalPacket, packet);
```

Comment:
```suggestion
            UsernameAndPassword auth = new();

            int unwrappedPacketLength = auth.Unwrap(packet, packetLength, out ProxyEndpoint remoteEndpoint);

            Assert.Equal(0, noAuth.WrapperLength);
            Assert.Null(remoteEndpoint);
            Assert.Equal(packetLength, unwrappedPacketLength);
            Assert.Equal(originalPacket, packet);
```

Same thing here, but the verification of the output `ProxyEndpoint` was also missing.